File: app/src/main/java/com/example/techfix/AppointmentsActivity.java
package com.example.techfix;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.techfix.data.AppointmentRepository;
import com.example.techfix.data.BranchRepository;
import com.example.techfix.data.FirestoreCallback;
import com.example.techfix.data.RepairServiceRepository;
import com.example.techfix.model.Appointment;
import com.example.techfix.model.Branch;
import com.example.techfix.model.RepairService;
import com.example.techfix.ui.AppointmentAdapter;
import com.example.techfix.ui.AppointmentDisplay;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AppointmentsActivity extends BaseActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_appointments);

        String userId = getIntent().getStringExtra("USER_ID");
        boolean filterCompleted = getIntent().getBooleanExtra("FILTER_COMPLETED", false);

        ((TextView) findViewById(R.id.tvAppointmentsTitle)).setText(filterCompleted ? "Repair History" : "My Appointments");

        RecyclerView recyclerView = findViewById(R.id.rvAppointments);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        TextView emptyState = findViewById(R.id.tvEmptyState);

        AppointmentRepository appointmentRepository = new AppointmentRepository();
        BranchRepository branchRepository = new BranchRepository();
        RepairServiceRepository serviceRepository = new RepairServiceRepository();

        appointmentRepository.getAppointmentsByCustomer(userId, new FirestoreCallback<Appointment>() {
            @Override
            public void onSuccess(List<Appointment> allAppointments) {
                List<Appointment> filtered = new ArrayList<>();
                for (Appointment a : allAppointments) {
                    if (!filterCompleted || "Completed".equals(a.status)) {
                        filtered.add(a);
                    }
                }

                if (filtered.isEmpty()) {
                    emptyState.setVisibility(View.VISIBLE);
                    emptyState.setText(filterCompleted ? "No completed repairs yet." : "No appointments yet.");
                    return;
                }
                emptyState.setVisibility(View.GONE);

                // Fetch branch and service reference data once, then join locally.
                branchRepository.getAllBranches(new FirestoreCallback<Branch>() {
                    @Override
                    public void onSuccess(List<Branch> branches) {
                        Map<String, String> branchNames = new HashMap<>();
                        for (Branch b : branches) branchNames.put(b.branchId, b.name);

                        serviceRepository.getAllServices(new FirestoreCallback<RepairService>() {
                            @Override
                            public void onSuccess(List<RepairService> services) {
                                Map<String, String> serviceNames = new HashMap<>();
                                for (RepairService s : services) serviceNames.put(s.serviceId, s.name);

                                List<AppointmentDisplay> displayList = new ArrayList<>();
                                for (Appointment a : filtered) {
                                    displayList.add(new AppointmentDisplay(
                                            a.appointmentId,
                                            a.requestedDate,
                                            branchNames.getOrDefault(a.branchId, "Unknown Branch"),
                                            serviceNames.getOrDefault(a.serviceId, "Unknown Service"),
                                            a.status
                                    ));
                                }

                                recyclerView.setAdapter(new AppointmentAdapter(displayList, appointment -> {
                                    Intent intent = new Intent(AppointmentsActivity.this, AppointmentDetailActivity.class);
                                    intent.putExtra("APPOINTMENT_ID", appointment.appointmentId);
                                    startActivity(intent);
                                }));
                            }

                            @Override
                            public void onFailure(Exception e) {}
                        });
                    }

                    @Override
                    public void onFailure(Exception e) {}
                });
            }

            @Override
            public void onFailure(Exception e) {}
        });
    }
}

File: app/src/main/java/com/example/techfix/AppointmentsActivity.kt
package com.example.techfix

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.techfix.data.AppDatabase
import com.example.techfix.ui.AppointmentAdapter
import com.example.techfix.ui.AppointmentDisplay
import kotlinx.coroutines.launch

class AppointmentsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_appointments)

        val userId = intent.getIntExtra("USER_ID", -1)
        val filterCompleted = intent.getBooleanExtra("FILTER_COMPLETED", false)

        findViewById<TextView>(R.id.tvAppointmentsTitle).text =
            if (filterCompleted) "Repair History" else "My Appointments"

        val recyclerView = findViewById<RecyclerView>(R.id.rvAppointments)
        recyclerView.layoutManager = LinearLayoutManager(this)
        val emptyState = findViewById<TextView>(R.id.tvEmptyState)

        val db = AppDatabase.getDatabase(applicationContext)

        lifecycleScope.launch {
            val allAppointments = db.appointmentDao().getAppointmentsByCustomer(userId)
            val filtered = if (filterCompleted) {
                allAppointments.filter { it.status == "Completed" }
            } else {
                allAppointments
            }

            if (filtered.isEmpty()) {
                emptyState.visibility = View.VISIBLE
                emptyState.text = if (filterCompleted) "No completed repairs yet." else "No appointments yet."
                return@launch
            }

            val displayList = filtered.map { appointment ->
                val branch = db.branchDao().getBranchById(appointment.branchId)
                val service = db.repairServiceDao().getServiceById(appointment.serviceId)
                AppointmentDisplay(
                    appointmentId = appointment.appointmentId,
                    date = appointment.requestedDate,
                    branchName = branch?.name ?: "Unknown Branch",
                    serviceName = service?.name ?: "Unknown Service",
                    status = appointment.status
                )
            }

            recyclerView.adapter = AppointmentAdapter(displayList)
        }
    }
}